planet/http.py

```diff
@@ -394,26 +399,80 @@ async def _send(self, request, stream=False) -> httpx.Response:
 
         return http_resp
 
-    @asynccontextmanager
-    async def stream(
-            self, method: str,
-            url: str) -> AsyncGenerator[models.StreamingResponse, None]:
-        """Submit a request and get the response as a stream context manager.
+    async def write(self,
+                    url: str,
+                    filename: Optional[str] = None,
+                    directory: Path = Path('.'),
+                    overwrite: bool = False,
+                    progress_bar: bool = False) -> Path:
```

Comment:
@jreiberkyle this is not the right level for these features. Let's allow the Session class to work in a simple space with no progress bars and no directories. Just requests and responses and bytes going in and out. If we do, I think we'll find it easier to refactor and maintain over time. Principle of least concern is what I'm invoking.

Concretely, let's give the download methods the responsibility of ensuring that directories exist and that we have a valid destination filename. Also, those methods should create progress bars and pass them to Session.write, which should just do `progress.update()`.

Reply:
At this time, retry is managed by `Session`. For `Session` to manage retry, it has to submit the original request for the download (so it can retry that request). When it submits the request, the response is not available to the client download function. So, the download function cannot access the `total` value in the response headers. There are two ways I can think of to allow download function to create the progress bar:
1. pass two callbacks to `Session.write`, one for update and one for setting the total
2. have the download function submit the original request to get the total value, then have `Session.write` submit the original request again as a part of the write process

I tend to think more than one callback is complex and so prefer option 2 even though it will be less efficient communication-wise, but I am not sure my aversion to multiple callbacks is warrented. @sgillies what do you think?

Reply:
It doesn't look like tqdm really supports setting the total after the progress bar has been initialized, so I am going to go with option 2: have the download function submit the original request to get the total, then have session.write submit it again to perform the download.

Reply:
@jreiberkyle option 2 is better, I agree. Can the double requests be avoided by making a decorator form of the retrier and re-trying not the HTTP requests, but `download_asset()`? Like this

```python
    @retry
    async def download_asset(...):
        ...
        async with self._session._client.stream, method='GET', url=location) as resp:
            content_length = resp.headers['Content-Length']
            dl_path = Path(directory, filename or parsed_from_content_disposition_header)
            dl_path.parent.mkdir(exist_ok=True, parents=True)
            return await self._session._write_response(
                resp,
                dl_path,
                overwrite=overwrite,
                progress_bar=tqdm(total=content_length, ...)
             )
```

Maybe for the next iteration? I think double requests are tolerable for now. Would it be best to make sure the first one gets closed/finalized before the second one is made? That's something to look into.

It also occurs to me that file sizes are in the order manifest.json file, so there's another way to get them. Also making another request, but only one extra per order.

Reply:
Decorating the client method has an example here: https://github.com/hynek/stamina#production-grade-retries-made-easy.

Reply:
agreed that we can generalize the retry function and make it work for retrying `download_asset()`. In that case, we may want to change `Session.request` so that it does not also retry (thereby squaring the effective MAX_RETRIES), and then change all the functions calling `Session.request` so they DO retry. At any rate, it would require some more thought. So, I went with option 2 - two calls.